internal/internal_worker_base.go

```diff
@@ -604,3 +709,255 @@ func (bw *baseWorker) Stop() {
+	// Here we periodically check if we should permit increasing the
+	// poller count further. We do this by comparing the number of ingested items in the
+	// current period with the number of ingested items in the previous period. If we
+	// are successfully ingesting more items, then it makes sense to allow scaling up.
+	// If we aren't, then we're probably limited by how fast we can process the tasks
+	// and it's not worth increasing the poller count further.
+	for {
+		select {
+		case <-ticker.C:
+			prh.newPeriod()
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (prh *pollScalerReportHandle) newPeriod() {
+	ingestedThisPeriod := prh.ingestedThisPeriod.Swap(0)
+	ingestedLastPeriod := prh.ingestedLastPeriod.Swap(ingestedThisPeriod)
+	prh.scaleUpAllowed.Store(float64(ingestedThisPeriod) >= float64(ingestedLastPeriod)*1.1)
+}
+
+func newPollerSemaphore(maxPermits int) *pollerSemaphore {
+	ps := &pollerSemaphore{
+		maxPermits: maxPermits,
+		permits:    0,
+		bs:         make(chan barrier, 1),
+	}
+	ps.bs <- make(barrier)
+	return ps
+}
+
+func (ps *pollerSemaphore) acquire(ctx context.Context) error {
+	for {
+		// Acquire barrier.
+		b := <-ps.bs
+		if ps.permits < ps.maxPermits {
+			ps.permits++
+			// Release barrier.
+			ps.bs <- b
+			return nil
+		}
+		// Release barrier.
+		ps.bs <- b
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-b:
+			continue
+		}
+	}
+}
+
+func (ps *pollerSemaphore) release() {
+	// Acquire barrier.
+	b := <-ps.bs
+	ps.permits--
+	// Release one waiter if there are any waiting.
+	select {
+	case b <- struct{}{}:
+	default:
+	}
+	// Release barrier.
+	ps.bs <- b
+}
+
+func (ps *pollerSemaphore) updatePermits(maxPermits int) {
+	// Acquire barrier.
+	b := <-ps.bs
+	ps.maxPermits = maxPermits
+	// Release barrier.
+	ps.bs <- b
+}
+
+func newScalableTaskPoller(
+	poller taskPoller, logger log.Logger, pollerBehavior PollerBehavior) scalableTaskPoller {
+	tw := scalableTaskPoller{
+		taskPoller: poller,
+	}
+	switch p := pollerBehavior.(type) {
+	case *pollerBehaviorAutoscaling:
+		tw.pollerCount = p.initialNumberOfPollers
+		tw.pollerSemaphore = newPollerSemaphore(p.initialNumberOfPollers)
+		tw.pollerAutoscalerReportHandle = newPollScalerReportHandle(pollScalerReportHandleOptions{
+			initialPollerCount: p.initialNumberOfPollers,
+			maxPollerCount:     p.maximumNumberOfPollers,
+			minPollerCount:     p.minimumNumberOfPollers,
+			logger:             logger,
+			scaleCallback: func(newTarget int) {
+				tw.pollerSemaphore.updatePermits(newTarget)
+			},
+		})
+	case *pollerBehaviorSimpleMaximum:
+		tw.pollerCount = p.maximumNumberOfPollers
+	}
+	return tw
+}
+
+// balance checks if the poller type is balanced with other poller types. The goal is to ensure that
+// at least one poller of each type is running before allowing any poller of the given type to increase.
+func (pb *pollerBalancer) balance(ctx context.Context, pollerType string) error {
```

Comment:
I think this could potentially be susceptible to some of the stuff fixed in https://github.com/temporalio/sdk-core/pull/941 -- but I'm not sure. Adding a test is probably worth it.

Reply:
Looking at the core issue, This code actually won't apply in the `SimpleMaximum` case. We didn't change any behaviour in the `SimpleMaximum ` case for Go